### ISSUE TITLE
implement a more ideal libdeno API

### ISSUE THREAD

**Issue by ry** [denoland/deno#1010]
```c
// JS API

namespace libdeno {
  // To check one's sanity: synchronously print to stdout or stderr.
  print(x: string, isErr?: boolean): void;

  // Shared with the C side. Set by deno_new().  Either side is free to write
  // to it at any time. Caller should provide synchronization for shared.
  const shared: Uint8Array;

  // Triggers deno_sync_cb in C.
  //
  // Caller owns zeroCopy, it is guarenteed to not be garbaged collected nor
  // moved until deno_release() is called in C.
  //
  // The returned buffer maps to the returned buffer of deno_sync_cb().
  sync(zeroCopy: null | Uint8Array): null | Uint8Array;


// C API

// shared is mapped to libdeno.shared. Either side is free to write to it at
// any time. Caller should provide synchronization for shared.
deno_isolate* deno_new(deno_buf shared, deno_sync_cb cb);

void deno_delete(deno_isolate* d);

// Calls into JS
int deno_run_js(deno_isolate* d, void* user_data, const char* js_filename,
    const char* js_source);

// Calls into JS. Runs pending promises.
void deno_run_microtasks(deno_isolate* d, void* user_data);

// deno_sync_cb is called everytime libdeno.sync() is called.
//
// If the zeroCopy parameter was null, zero_copy_id will be 0 and zero_copy
// will have a null pointer. If zeroCopy was not null, zero_copy_id will be
// a positive number. The zero_copy buf can be accessed until deno_release()
// is called.
//
// The returned value can be an empty buf, in which case libdeno.sync will
// return null.
//
// If a return buf was supplied, its memory is now owned by V8. The caller must
// not free or otherwise access it again.
deno_buf (*deno_sync_cb)(void* user_data, int32_t zero_copy_id, deno_buf zero_copy);

// Releases a buffer that was supplied in deno_sync_cb. Thread safe.
void deno_release(int32_t zero_copy_id);
```

**Comment by ry** [comment]
Partially implemented in  0501330607b000e0913994e633b76410e1fd162c and 61cda728816b5916180f22d977ba741e2fadc8d9

**Comment by bartlomieju** [comment]
CC @ry, is this issue still relevant?
